README
Haskell bindings for the [Blurhash](https://github.com/woltapp/blurhash)
algorithm.

# Usage

```
% blurhash encode images/cat.jpg 3 3
KYLDf7IA~p^+x]S5xvW=M|
% blurhash decode "KYLDf7IA~p^+x]S5xvW=M|" 1199 1198 -o out.png
Saved image to out.png
```

# TODO
  implement this and see how close to blurhash you can get it
  https://www.mathworks.com/help/images/discrete-cosine-transform.html


CHANGELOG.md
# Revision history for blur

## 0.1.0.0 -- 2022-05-16

* First version. Released upon an unsuspecting world.


test.hs
import Blurhash.Encode
import Codec.Picture
import System.Environment

main = do
  (nx:ny:imgPath:_) <- getArgs
  Right dynImg <- readImage imgPath -- Either String DynamicImage
  let img = convertRGB8 dynImg
  let (w, h) = (fromIntegral $ imageWidth img, fromIntegral $ imageHeight img)
  let (x, y) = (read nx, read ny) :: (Int, Int)

  -- cat.jpg should get us "KYLDf7IA~p^+x]S5xvW=M|" for (3, 3)
  -- cat.jpg should get us "ARLDf7Di~W%M" for (2, 2)
  putStrLn $ "Test program!"
  putStrLn $ "Generating blurhash for " ++ show w ++ "x" ++ show h ++ " image..."
  putStrLn . blurhash (x, y) $ img



Gray.hs
import Codec.Picture
import System.Environment
import Math.FFT
import Data.Array.CArray

type Array2D = CArray (Int, Int) Double

grayscale :: PixelRGB8 -> Pixel8
grayscale (PixelRGB8 r g b) =
  linearToSrgb $ (0.2126 * r' + 0.7152 * g' + 0.0722 * b') / 255.0
  where
    r' = fromIntegral r
    g' = fromIntegral g
    b' = fromIntegral b

srgbToLinear :: Double -> Double
srgbToLinear value = ((x' + 0.055) / 1.055) ** 2.4
  where
    x = value / 255.0
    x' = if x <= 0.04045
         then x / 12.92
         else x

linearToSrgb :: Double -> Pixel8
linearToSrgb value = floor ret
  where
    x   = limit 0 1 value
    ret = if x <= 0.0031308
          then x * 12.92 * 255 + 0.5
          else (1.055 * (x ** (1.0 / 2.4)) - 0.055) * 255 + 0.5

limit low high x | x < low    = low
                 | x > high   = high
                 | otherwise  = x

oneDim :: (Pixel p) => Image p -> [p]
oneDim img = [pixelAt img i j | i <- [0..w], j <- [0..h]]
  where
    h = (imageHeight img) - 1
    w = (imageWidth img) - 1

toCArray :: Image Pixel8 -> Array2D
toCArray img@(Image w h _) =
  listArray ((0, 0), (w - 1, h - 1)) .
  map (fromIntegral . toInteger) .
  oneDim $ img

pixelR, pixelG, pixelB :: PixelRGB8 -> Pixel8
pixelR (PixelRGB8 r _ _) = r
pixelG (PixelRGB8 _ g _) = g
pixelB (PixelRGB8 _ _ b) = b

extractChannel :: (PixelRGB8 -> Pixel8) -> Image PixelRGB8 -> Array2D
extractChannel f img@(Image w h _) =
  listArray ((0, 0), (w - 1, h - 1)) .
  map (fromIntegral . toInteger . f) .
  oneDim $ img

channelR = extractChannel pixelR
channelG = extractChannel pixelG
channelB = extractChannel pixelB

main = do
  (imgPath:_)  <- getArgs
  Right dynImg <- readImage imgPath
  let img = convertRGB8 dynImg

  let gray = pixelMap grayscale img
  let red = pixelMap (\(PixelRGB8 r g b) -> PixelRGB8 r 0 0) img
  let green = pixelMap (\(PixelRGB8 r g b) -> PixelRGB8 0 g 0) img
  let blue = pixelMap (\(PixelRGB8 r g b) -> PixelRGB8 0 0 b) img

  let fft = dht $ toCArray gray :: Array2D
  let (w, h) = (imageWidth img, imageHeight img)
  let fft_img = generateImage (\x y -> linearToSrgb $ fft ! (x, y)) w h

  saveJpgImage 100 "red_cat.jpg" $ ImageRGB8 red
  saveJpgImage 100 "green_cat.jpg" $ ImageRGB8 green
  saveJpgImage 100 "blue_cat.jpg" $ ImageRGB8 blue
  saveJpgImage 100 "dct_cat.jpg" $ ImageY8 fft_img



Main.hs
import Blurhash.Encode
import Blurhash.Decode
import Codec.Picture
import System.Environment
import Options.Applicative

data Command = Encode String Int Int
             | Decode String Int Int String

encode :: Parser Command
encode = Encode
     <$> strArgument
        (  metavar "BLURHASH"
        <> help "Blurhash to be decoded into a PNG")
     <*> xComponents
     <*> yComponents

decode :: Parser Command
decode = Decode
     <$> strArgument
        (  metavar "IMAGE_PATH"
        <> help "Path to image to be encoded into a blurhash")
     <*> width
     <*> height
     <*> output

xComponents :: Parser Int
xComponents = option auto
             ( long "nx"
            <> short 'x'
            <> metavar "NX"
            <> value 3
            <> help "Number of x components (1-9) (default: 3)")

yComponents :: Parser Int
yComponents = option auto
             ( long "ny"
            <> short 'y'
            <> metavar "NY"
            <> value 3
            <> help "Number of y components (1-9) (default: 3)")

width :: Parser Int
width = argument auto
             ( metavar "W"
            <> help "Width in pixels")

height :: Parser Int
height = argument auto
             ( metavar "H"
            <> help "Height in pixels")

output :: Parser String
output = strOption
             ( long "out"
            <> short 'o'
            <> metavar "FILE"
            <> help "Output file for the generated PNG")

encOptions :: Mod CommandFields Command
encOptions =  command "encode" (info encode (progDesc "Encode an image into a hash"))

decOptions :: Mod CommandFields Command
decOptions =  command "decode" (info decode (progDesc "Decode a hash into an image"))

optsParser =
  info
    (helper <*> versionOption <*> programOptions)
    (fullDesc <> progDesc "blurhash command [arguments...] [flags...]"
              <> header "blurhash: a small program to encode images into blurhashes and decode hashes into blurred images")

versionOption = infoOption "0.1" (long "version" <> short 'v' <> help "Show Version")

programOptions :: Parser Command
programOptions = hsubparser (encOptions <> decOptions)

doEncode imgPath x y = do
  Right dynImg <- readImage imgPath -- Either String DynamicImage
  let img = convertRGB8 dynImg
  let (w, h) = (fromIntegral $ imageWidth img, fromIntegral $ imageHeight img)

  putStrLn . blurEncode (x, y) $ img

doDecode hash w h o = do
  let img = blurDecode hash (w, h)
  savePngImage o $ ImageRGB8 img
  putStrLn $ "Saved image to " ++ o

main = do
  opts <- execParser optsParser
  case opts of
    Encode file x y   -> doEncode file x y
    Decode hash w h o -> doDecode hash w h o



Blurhash/Decode.hs
module Blurhash.Decode (blurDecode) where

import System.Environment
import Codec.Picture
import Data.List
import Data.Maybe
import Data.Bits

import Blurhash.Common

unhash :: String -> Int -> Int -> Int
unhash hash start end = unhash83 slice
  where
    slice = take (end - start) . drop start $ hash

unhash83 :: String -> Int
unhash83 hash = total
  where
    cipher = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~"

    -- this is a list of Maybes
    elems = reverse . catMaybes $ map (\h -> elemIndex h cipher) hash
    total = fst $ foldl (\(s, n) v -> (s + v * 83 ^ n, n + 1)) (0, 0) elems

numComponents :: String -> (Int, Int)
numComponents hash = (nx, ny)
  where
    size = unhash hash 0 1
    ny = (size `div`  9) + 1
    nx = (size `mod` 9) + 1

maxValue :: String -> Double
maxValue hash = (fromIntegral qnt + 1) / 166
  where
    qnt = unhash hash 1 2

decodeDC :: String -> RGB
decodeDC hash = (srgbToLinear r, srgbToLinear g, srgbToLinear b)
  where
    avgColor = unhash hash 2 6
    r = fromIntegral $ avgColor `shift` (-16)
    g = fromIntegral $ (avgColor `shift` (-8)) .&. 255
    b = fromIntegral $ avgColor .&. 255

decodeAC :: Double -> Int -> RGB
decodeAC maxV ac = (r', g', b')
  where
    r  = fromIntegral $ ac `div` (19 * 19)
    g  = fromIntegral $ (ac `div` 19) `mod` 19
    b  = fromIntegral $ ac `mod` 19
    r' = (signpow ((r - 9) / 9) 2.0) * maxV
    g' = (signpow ((g - 9) / 9) 2.0) * maxV
    b' = (signpow ((b - 9) / 9) 2.0) * maxV

decodeACs:: String -> [RGB]
decodeACs hash = (decodeDC hash) : acs
  where
    (nx, ny) = numComponents hash

    -- this would normally go from 0..(nx * ny) - 1
    -- but since we already have the first component from `decodeDC`, we
    -- only need 1..(nx * ny) - 1
    acValues = [unhash hash (4 + x * 2) (6 + x * 2) | x <- [1..(nx * ny) - 1]]
    acs = map (decodeAC (maxValue hash)) acValues

blurDecode :: String -> (Int, Int) -> Image PixelRGB8
blurDecode hash (w, h) = generateImage (makePixel acs (w, h) (nx, ny)) w h
  where
    (nx, ny) = numComponents hash
    acs = decodeACs hash

makePixel :: [RGB] -> (Int, Int) -> (Int, Int) -> Int -> Int -> PixelRGB8
makePixel acs (w, h) (nx, ny) x y = PixelRGB8 r' g' b'
  where
    basis i j = (cos (pi * i' * x' / w')) * (cos (pi * j' * y' / h'))
      where
        i' = fromIntegral i
        j' = fromIntegral j
        x' = fromIntegral x
        y' = fromIntegral y
        w' = fromIntegral w
        h' = fromIntegral h
    bases = [basis i j | j <- [0..ny - 1], i <- [0..nx - 1]]

    reds   = map (\(r, _, _) -> r) acs
    greens = map (\(_, g, _) -> g) acs
    blues  = map (\(_, _, b) -> b) acs

    r = sum $ zipWith (*) bases reds
    g = sum $ zipWith (*) bases greens
    b = sum $ zipWith (*) bases blues

    r' = fromIntegral . limit 0 255 $ linearToSrgb r
    g' = fromIntegral . limit 0 255 $ linearToSrgb g
    b' = fromIntegral . limit 0 255 $ linearToSrgb b

main = do
  (hash:w:h:file:[]) <- getArgs
  let (width, height) = (read w, read h) :: (Int, Int)

  let img = blurDecode hash (width, height)
  savePngImage file $ ImageRGB8 img
  putStrLn $ "Saved image to " ++ file



Blurhash/Encode.hs
module Blurhash.Encode (blurEncode) where

import Math.FFT
import Codec.Picture
import System.Environment
import Control.Applicative
import Data.Bits

import Blurhash.Common

oneDim :: (Pixel p) => Image p -> [p]
oneDim img = [pixelAt img i j | j <- [0..h], i <- [0..w]]
  where
    h = (imageHeight img) - 1
    w = (imageWidth img) - 1

pixelR, pixelG, pixelB :: PixelRGB8 -> Pixel8
pixelR (PixelRGB8 r _ _) = r
pixelG (PixelRGB8 _ g _) = g
pixelB (PixelRGB8 _ _ b) = b

extractChannel :: (PixelRGB8 -> Pixel8) -> Image PixelRGB8 -> [Pixel8]
extractChannel f img@(Image w h _) = map f . oneDim $ img

channelR = extractChannel pixelR
channelG = extractChannel pixelG
channelB = extractChannel pixelB

manualDCT :: Image PixelRGB8 -> (Int, Int) -> [RGB]
manualDCT img (nx, ny) = [basisFunction img i j | j <- [0..ny - 1], i <- [0..nx - 1]]

basisFunction :: Image PixelRGB8 -> Int -> Int -> RGB
basisFunction img i j = (r, g, b)
  where
    (w, h) = (fromIntegral $ imageWidth img, fromIntegral $ imageHeight img)
    (i', j') = (fromIntegral i, fromIntegral j)

    -- not sure why `x` and `y` are Doubles, but I'm not mad about it
    basis x y = (cos (pi * i' * x / w)) * (cos (pi * j' * y / h))
    bases = [basis x y | y <- [0..h - 1], x <- [0..w - 1]] :: [Double]

    red   = zipWith (*) bases . map (srgbToLinear . fromIntegral) . channelR $ img
    green = zipWith (*) bases . map (srgbToLinear . fromIntegral) . channelG $ img
    blue  = zipWith (*) bases . map (srgbToLinear . fromIntegral) . channelB $ img

    norm   = if i == 0 && j == 0
             then 1.0
             else 2.0
    scale = norm / (w * h)
    r = scale * (sum red)
    g = scale * (sum green)
    b = scale * (sum blue)

blurEncode :: (Int, Int) -> Image PixelRGB8 -> String
blurEncode (nx, ny) img =
  (hash83 1 nc) ++ (hash83 1 qntAC) ++ (hash83 4 dc) ++ (foldr (\ac str -> (hash83 2 ac) ++ str) "" comps)
  where
    nc    = (nx - 1) + (ny - 1) * 9
    num_c = nx * ny

    dct   = take num_c . manualDCT img $ (nx, ny)
    acs   = drop 1 dct

    -- average color
    dc    = encodeDC $ head dct

    maxAC = foldl (\m (a, b, c) -> maximum $ map abs [m, a, b, c]) 0.0 acs

    qntAC  = limit 0 82 . floor $ maxAC * 166 - 0.5 :: Int
    maxAC' = ((fromIntegral qntAC) + 1) / 166 :: Double

    comps = map (encodeAC maxAC') acs

encodeDC :: RGB -> Int
encodeDC (r, g, b) = fromIntegral . toInteger $ (r' `shift` 16) + (g' `shift` 8) + b'
  where
    r' = linearToSrgb $ r :: Int
    g' = linearToSrgb $ g :: Int
    b' = linearToSrgb $ b :: Int

encodeAC :: Double -> RGB -> Int
encodeAC maxV (r, g, b) = r' * (19 * 19) + g' * 19 + b'
  where
    r' = limit 0 18 . floor $ (signpow (r / maxV) 0.5) * 9 + 9.5
    g' = limit 0 18 . floor $ (signpow (g / maxV) 0.5) * 9 + 9.5
    b' = limit 0 18 . floor $ (signpow (b / maxV) 0.5) * 9 + 9.5


hash83 :: Int -> Int -> String
hash83 places value = hash83' 1 places value

hash83' :: Int -> Int -> Int -> String
hash83' cur places value 
  | cur > places = ""
  | otherwise     = res : (hash83' (cur + 1) places value)
    where
      cipher = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~"
      digit  = (value `div` (83 ^ (places - cur))) `mod` 83
      res    = cipher !! digit

main = do
  (nx:ny:imgPath:_) <- getArgs
  Right dynImg <- readImage imgPath -- Either String DynamicImage
  let img = convertRGB8 dynImg
  let (w, h) = (fromIntegral $ imageWidth img, fromIntegral $ imageHeight img)
  let (x, y) = (read nx, read ny) :: (Int, Int)

  -- cat.jpg should get us "KYLDf7IA~p^+x]S5xvW=M|" for (3, 3)
  -- cat.jpg should get us "ARLDf7Di~W%M" for (2, 2)
  putStrLn $ "Generating blurhash for " ++ show w ++ "x" ++ show h ++ " image..."
  putStrLn . blurEncode (x, y) $ img

